File: convert_to_dwc.r
#!/usr/bin/env Rscript
# Transform data to Darwin Core Archive
#
# Schema reference: https://example.org/data-sample
# Event hierarchy:
#   - Cruise events (deployment/recovery) with footprintWKT
#   - Station events (points) as children

library(tidyverse)
library(readxl)

# Read source data from Excel
station_df <- read_excel("Data_sample.xlsx", sheet = "Station")
cpue_df <- read_excel("Data_sample.xlsx", sheet = "CPUE")
measurements_df <- read_excel("Data_sample.xlsx", sheet = "Measurements")

cat("Loaded data:\n")
cat("  Station:", nrow(station_df), "rows\n")
cat("  CPUE:", nrow(cpue_df), "rows\n")
cat("  Measurements:", nrow(measurements_df), "rows\n")

# ============================================================================
# Create parent Cruise events (deployment and recovery)
# Schema: Cruise events have footprintWKT from all stations
# ============================================================================

cruise_events <- station_df %>%
  group_by(cruise_id, type) %>%
  summarise(
    eventDate = first(datetime),
    # Build LINESTRING from all start and end coordinates
    coords = paste(
      na.omit(paste(lon_start, lat_start)),
      na.omit(paste(lon_end, lat_end)),
      collapse = ", "
    ),
    .groups = "drop"
  ) %>%
  mutate(
    eventID = paste0(cruise_id, "_", type),
    eventType = paste(type, "cruise"),
    parentEventID = NA_character_,
    footprintWKT = paste0("LINESTRING (", coords, ")")
  ) %>%
  select(eventID, eventDate, eventType, parentEventID, footprintWKT)

cat("\nCreated", nrow(cruise_events), "parent cruise events\n")

# ============================================================================
# Transform Station -> Event core (as point locations)
# Schema class: Station -> dwc:Event
# Schema slot: station -> dwc:eventID and dwc:locationID
# ============================================================================

station_events <- station_df %>%
  mutate(
    # Create parentEventID based on cruise_id and type
    parentEventID = paste0(cruise_id, "_", type)
  ) %>%
  transmute(
    eventID = station,                    # Schema slot: station -> dwc:eventID
    locationID = station,                 # Schema slot: station -> dwc:locationID
    parentEventID = parentEventID,
    eventDate = datetime,                 # Schema slot: datetime -> dwc:eventDate
    eventType = type,                     # Schema slot: type -> dwc:eventType
    decimalLatitude = lat_start,          # Schema slot: lat_start -> dwc:decimalLatitude (point)
    decimalLongitude = lon_start,         # Schema slot: lon_start -> dwc:decimalLongitude (point)
    minimumDepthInMeters = depth,         # Schema slot: depth -> dwc:minimumDepthInMeters
    maximumDepthInMeters = depth,         # Schema slot: depth -> dwc:maximumDepthInMeters
    eventRemarks = notes,                 # Schema slot: notes -> dwc:eventRemarks
    recordedBy = participants             # Schema slot: participants -> dwc:recordedBy
  )

cat("Created", nrow(station_events), "station events\n")

# Combine cruise and station events
event_core <- bind_rows(cruise_events, station_events)

cat("Total events in core:", nrow(event_core), "\n")

# ============================================================================
# Transform CPUE -> Occurrence extension
# Schema class: CPUE -> dwc:Occurrence
# ============================================================================

occurrence_ext <- cpue_df %>%
  mutate(
    occurrenceID = paste(Station, Pot_ID, Species, row_number(), sep = "_")
  ) %>%
  transmute(
    occurrenceID = occurrenceID,
    eventID = Station,
    vernacularName = Species,
    individualCount = Catch,
    basisOfRecord = "HumanObservation",
    occurrenceRemarks = Notes
  )

cat("\nCreated Occurrence extension from CPUE with", nrow(occurrence_ext), "occurrences\n")

# ============================================================================
# Transform Measurements -> Occurrence extension
# Schema class: Measurements -> dwc:Occurrence
# ============================================================================

measurement_occurrences <- measurements_df %>%
  mutate(
    occurrenceID = paste("MEAS", Station, Species, row_number(), sep = "_"),
    # Combine barotrauma and notes
    occurrenceRemarks = case_when(
      !is.na(Barotrauma) & !is.na(Notes) ~ paste0("Barotrauma: ", Barotrauma, "; ", Notes),
      !is.na(Barotrauma) ~ paste0("Barotrauma: ", Barotrauma),
      !is.na(Notes) ~ Notes,
      TRUE ~ NA_character_
    )
  ) %>%
  transmute(
    occurrenceID = occurrenceID,
    eventID = Station,
    vernacularName = Species,
    sex = Sex,
    basisOfRecord = "HumanObservation",
    occurrenceRemarks = occurrenceRemarks
  )

cat("Created Occurrence extension from Measurements with", nrow(measurement_occurrences), "occurrences\n")

# Combine occurrences
occurrence_combined <- bind_rows(occurrence_ext, measurement_occurrences)

cat("Combined occurrences:", nrow(occurrence_combined), "total\n")

# ============================================================================
# Create MeasurementOrFact extension
# ============================================================================

# Organism measurements from Measurements sheet
measurements_with_id <- measurements_df %>%
  mutate(
    occurrenceID = paste("MEAS", Station, Species, row_number(), sep = "_")
  )

# Total length
length_mof <- measurements_with_id %>%
  filter(!is.na(TL_mm)) %>%
  transmute(
    occurrenceID = occurrenceID,
    measurementType = "total length",
    measurementValue = as.character(TL_mm),
    measurementUnit = "mm",
    measurementUnitID = "http://qudt.org/vocab/unit/MilliM"
  )

# Weight (recorded)
weight_recorded_mof <- measurements_with_id %>%
  filter(!is.na(Wt_g_recorded)) %>%
  transmute(
    occurrenceID = occurrenceID,
    measurementType = "weight (recorded)",
    measurementValue = as.character(Wt_g_recorded),
    measurementUnit = "g",
    measurementUnitID = "http://qudt.org/vocab/unit/GM"
  )

# Scale tare weight
tare_mof <- measurements_with_id %>%
  filter(!is.na(scale_tare_g)) %>%
  transmute(
    occurrenceID = occurrenceID,
    measurementType = "scale tare weight",
    measurementValue = as.character(scale_tare_g),
    measurementUnit = "g",
    measurementUnitID = "http://qudt.org/vocab/unit/GM"
  )

# Weight (calculated)
weight_mof <- measurements_with_id %>%
  filter(!is.na(Wt_g)) %>%
  transmute(
    occurrenceID = occurrenceID,
    measurementType = "weight",
    measurementValue = as.character(Wt_g),
    measurementUnit = "g",
    measurementUnitID = "http://qudt.org/vocab/unit/GM"
  )

# Retained status as measurement
# Schema slot: retained -> dwc:measurementValue (transformation_type: pivot_to_mof)
retained_mof <- measurements_with_id %>%
  filter(!is.na(Retained)) %>%
  transmute(
    occurrenceID = occurrenceID,
    measurementType = "retained",
    measurementValue = as.character(Retained),
    measurementUnit = NA_character_
  )

# Environmental measurements from Station sheet
# Wind speed
wind_speed_mof <- station_df %>%
  filter(!is.na(wind_speed)) %>%
  transmute(
    eventID = station,
    measurementType = "wind speed",
    measurementValue = as.character(wind_speed),
    measurementUnit = "kn",
    measurementUnitID = "http://qudt.org/vocab/unit/KN"
  )

# Wind direction
wind_dir_mof <- station_df %>%
  filter(!is.na(wind_dir)) %>%
  transmute(
    eventID = station,
    measurementType = "wind direction",
    measurementValue = as.character(wind_dir),
    measurementUnit = NA_character_
  )

# Wave height
wave_height_mof <- station_df %>%
  filter(!is.na(wave_height)) %>%
  transmute(
    eventID = station,
    measurementType = "wave height",
    measurementValue = as.character(wave_height),
    measurementUnit = NA_character_
  )

# Cloud cover
cloud_cover_mof <- station_df %>%
  filter(!is.na(cloud_cover_10th)) %>%
  transmute(
    eventID = station,
    measurementType = "cloud cover",
    measurementValue = as.character(cloud_cover_10th),
    measurementUnit = "tenths"
  )

# Ropeless gear ID
ropeless_mof <- station_df %>%
  filter(!is.na(ropeless_id)) %>%
  transmute(
    eventID = station,
    measurementType = "ropeless gear ID",
    measurementValue = as.character(ropeless_id),
    measurementUnit = NA_character_
  )

# Cruise ID as measurement
# Schema slot: cruise_id -> dwc:measurementValue (transformation_type: pivot_to_mof)
cruise_id_mof <- station_df %>%
  filter(!is.na(cruise_id)) %>%
  transmute(
    eventID = station,
    measurementType = "cruise ID",
    measurementValue = as.character(cruise_id),
    measurementUnit = NA_character_
  )

# Gear measurements from CPUE
# Pot position
pot_position_mof <- cpue_df %>%
  filter(!is.na(Pot_position)) %>%
  transmute(
    eventID = Station,
    measurementType = "pot position",
    measurementValue = as.character(Pot_position),
    measurementUnit = NA_character_
  )

# Pot ID
pot_id_mof <- cpue_df %>%
  filter(!is.na(Pot_ID)) %>%
  transmute(
    eventID = Station,
    measurementType = "pot ID",
    measurementValue = as.character(Pot_ID),
    measurementUnit = NA_character_
  )

# Distance category (near/far) from CPUE
near_far_cpue_mof <- cpue_df %>%
  filter(!is.na(Near_Far)) %>%
  transmute(
    eventID = Station,
    measurementType = "distance category",
    measurementValue = as.character(Near_Far),
    measurementUnit = NA_character_
  )

# Distance category from Measurements
near_far_meas_mof <- measurements_df %>%
  filter(!is.na(`Near/Far`)) %>%
  transmute(
    eventID = Station,
    measurementType = "distance category",
    measurementValue = as.character(`Near/Far`),
    measurementUnit = NA_character_
  )

# Combine all measurements
measurement_or_fact <- bind_rows(
  length_mof,
  weight_recorded_mof,
  tare_mof,
  weight_mof,
  retained_mof,
  wind_speed_mof,
  wind_dir_mof,
  wave_height_mof,
  cloud_cover_mof,
  ropeless_mof,
  cruise_id_mof,
  pot_position_mof,
  pot_id_mof,
  near_far_cpue_mof,
  near_far_meas_mof
)

cat("\nCreated MeasurementOrFact extension with", nrow(measurement_or_fact), "measurements\n")

# ============================================================================
# Write outputs
# ============================================================================

write_csv(event_core, "outputs/r_dwc_event.csv", na = "")
write_csv(occurrence_combined, "outputs/r_dwc_occurrence.csv", na = "")
write_csv(measurement_or_fact, "outputs/r_dwc_measurementorfact.csv", na = "")

File: convert_to_dwc.py
#!/usr/bin/env python3
"""
Transform data to Darwin Core Archive

Schema reference: https://example.org/data-sample
Event hierarchy: 
  - Cruise events (deployment/recovery) with footprintWKT
  - Station events (points) as children
"""

import pandas as pd
import numpy as np

# Read source data from Excel
station_df = pd.read_excel('Data_sample.xlsx', sheet_name='Station')
cpue_df = pd.read_excel('Data_sample.xlsx', sheet_name='CPUE')
measurements_df = pd.read_excel('Data_sample.xlsx', sheet_name='Measurements')

print(f"Loaded data:")
print(f"  Station: {len(station_df)} rows")
print(f"  CPUE: {len(cpue_df)} rows")
print(f"  Measurements: {len(measurements_df)} rows")

# ============================================================================
# Create parent Cruise events (deployment and recovery)
# Schema: Cruise events have footprintWKT from all stations
# ============================================================================

# Get unique cruise_id and types
cruise_types = station_df.groupby(['cruise_id', 'type']).agg({
    'lon_start': lambda x: list(x),
    'lat_start': lambda x: list(x),
    'lon_end': lambda x: list(x),
    'lat_end': lambda x: list(x),
    'datetime': 'first'
}).reset_index()

cruise_events = []
for idx, row in cruise_types.iterrows():
    # Build LINESTRING from all station start/end points for this cruise+type
    coords = []
    for lon_s, lat_s, lon_e, lat_e in zip(row['lon_start'], row['lat_start'], 
                                            row['lon_end'], row['lat_end']):
        if pd.notna(lon_s) and pd.notna(lat_s):
            coords.append(f"{lon_s} {lat_s}")
        if pd.notna(lon_e) and pd.notna(lat_e):
            coords.append(f"{lon_e} {lat_e}")
    
    footprint = f"LINESTRING ({', '.join(coords)})" if coords else None
    
    cruise_events.append({
        'eventID': f"{int(row['cruise_id'])}_{row['type']}",
        'eventDate': row['datetime'],
        'eventType': f"{row['type']} cruise",
        'parentEventID': None,
        'footprintWKT': footprint
    })

cruise_event_df = pd.DataFrame(cruise_events)

print(f"\nCreated {len(cruise_event_df)} parent cruise events")

# ============================================================================
# Transform Station -> Event core (as point locations)
# Schema class: Station -> dwc:Event
# Schema slot: station -> dwc:eventID and dwc:locationID
# ============================================================================

station_events = station_df.copy()

# Create parentEventID based on cruise_id and type
station_events['parentEventID'] = (station_events['cruise_id'].astype(int).astype(str) + '_' + 
                                    station_events['type'])

station_events = station_events.rename(columns={
    'station': 'eventID',           # Schema slot: station -> dwc:eventID
    'datetime': 'eventDate',        # Schema slot: datetime -> dwc:eventDate
    'type': 'eventType',            # Schema slot: type -> dwc:eventType
    'lat_start': 'decimalLatitude', # Schema slot: lat_start -> dwc:decimalLatitude (point)
    'lon_start': 'decimalLongitude',# Schema slot: lon_start -> dwc:decimalLongitude (point)
    'depth': 'minimumDepthInMeters',# Schema slot: depth -> dwc:minimumDepthInMeters
    'notes': 'eventRemarks',        # Schema slot: notes -> dwc:eventRemarks
    'participants': 'recordedBy'    # Schema slot: participants -> dwc:recordedBy
})

# Add locationID (same as eventID for stations)
# Schema slot: station -> dwc:locationID
station_events['locationID'] = station_events['eventID']

# Add maximumDepthInMeters (same as minimum for single measurements)
station_events['maximumDepthInMeters'] = station_events['minimumDepthInMeters']

# Select final columns for Station events
station_events = station_events[[
    'eventID', 'locationID', 'parentEventID', 'eventDate', 'eventType', 
    'decimalLatitude', 'decimalLongitude', 'minimumDepthInMeters', 
    'maximumDepthInMeters', 'eventRemarks', 'recordedBy'
]]

# Combine cruise and station events
event_core = pd.concat([cruise_event_df, station_events], ignore_index=True)

print(f"Created {len(station_events)} station events")
print(f"Total events in core: {len(event_core)}")

# ============================================================================
# Transform CPUE -> Occurrence extension
# Schema class: CPUE -> dwc:Occurrence
# ============================================================================

cpue_df['occurrenceID'] = (cpue_df['Station'].astype(str) + '_' + 
                            cpue_df['Pot_ID'].astype(str) + '_' + 
                            cpue_df['Species'].astype(str) + '_' + 
                            (cpue_df.index + 1).astype(str))

occurrence_ext = cpue_df[['occurrenceID', 'Station', 'Species', 'Catch', 'Notes']].copy()

occurrence_ext = occurrence_ext.rename(columns={
    'Station': 'eventID',
    'Species': 'vernacularName',
    'Catch': 'individualCount',
    'Notes': 'occurrenceRemarks'
})

occurrence_ext['basisOfRecord'] = 'HumanObservation'

print(f"\nCreated Occurrence extension from CPUE with {len(occurrence_ext)} occurrences")

# ============================================================================
# Transform Measurements -> Occurrence extension
# Schema class: Measurements -> dwc:Occurrence
# ============================================================================

measurements_df['occurrenceID'] = ('MEAS_' + 
                                   measurements_df['Station'].astype(str) + '_' + 
                                   measurements_df['Species'].astype(str) + '_' + 
                                   (measurements_df.index + 1).astype(str))

# Combine barotrauma and notes
def combine_remarks(row):
    parts = []
    if pd.notna(row.get('Barotrauma')):
        parts.append(f"Barotrauma: {row['Barotrauma']}")
    if pd.notna(row.get('Notes')):
        parts.append(str(row['Notes']))
    return '; '.join(parts) if parts else None

measurements_df['combined_remarks'] = measurements_df.apply(combine_remarks, axis=1)

measurement_occurrences = measurements_df[['occurrenceID', 'Station', 'Species', 
                                           'Sex', 'combined_remarks']].copy()

measurement_occurrences = measurement_occurrences.rename(columns={
    'Station': 'eventID',
    'Species': 'vernacularName',
    'Sex': 'sex',
    'combined_remarks': 'occurrenceRemarks'
})

measurement_occurrences['basisOfRecord'] = 'HumanObservation'

print(f"Created Occurrence extension from Measurements with {len(measurement_occurrences)} occurrences")

# Combine occurrences
occurrence_combined = pd.concat([occurrence_ext, measurement_occurrences], ignore_index=True)

print(f"Combined occurrences: {len(occurrence_combined)} total")

# ============================================================================
# Create MeasurementOrFact extension
# ============================================================================

measurement_records = []

# Organism measurements from Measurements sheet
for idx, row in measurements_df.iterrows():
    occ_id = row['occurrenceID']
    
    # Total length
    if pd.notna(row['TL_mm']):
        measurement_records.append({
            'occurrenceID': occ_id,
            'measurementType': 'total length',
            'measurementValue': str(row['TL_mm']),
            'measurementUnit': 'mm',
            'measurementUnitID': 'http://qudt.org/vocab/unit/MilliM'
        })
    
    # Weight (recorded)
    if pd.notna(row['Wt_g_recorded']):
        measurement_records.append({
            'occurrenceID': occ_id,
            'measurementType': 'weight (recorded)',
            'measurementValue': str(row['Wt_g_recorded']),
            'measurementUnit': 'g',
            'measurementUnitID': 'http://qudt.org/vocab/unit/GM'
        })
    
    # Scale tare weight
    if pd.notna(row['scale_tare_g']):
        measurement_records.append({
            'occurrenceID': occ_id,
            'measurementType': 'scale tare weight',
            'measurementValue': str(row['scale_tare_g']),
            'measurementUnit': 'g',
            'measurementUnitID': 'http://qudt.org/vocab/unit/GM'
        })
    
    # Weight (calculated)
    if pd.notna(row['Wt_g']):
        measurement_records.append({
            'occurrenceID': occ_id,
            'measurementType': 'weight',
            'measurementValue': str(row['Wt_g']),
            'measurementUnit': 'g',
            'measurementUnitID': 'http://qudt.org/vocab/unit/GM'
        })
    
    # Retained status as measurement
    # Schema slot: retained -> dwc:measurementValue (transformation_type: pivot_to_mof)
    if pd.notna(row['Retained']):
        measurement_records.append({
            'occurrenceID': occ_id,
            'measurementType': 'retained',
            'measurementValue': str(row['Retained']),
            'measurementUnit': ''
        })

# Environmental measurements from Station sheet
for idx, row in station_df.iterrows():
    event_id = row['station']
    
    # Wind speed
    if pd.notna(row['wind_speed']):
        measurement_records.append({
            'eventID': event_id,
            'measurementType': 'wind speed',
            'measurementValue': str(row['wind_speed']),
            'measurementUnit': 'kn',
            'measurementUnitID': 'http://qudt.org/vocab/unit/KN'
        })
    
    # Wind direction
    if pd.notna(row['wind_dir']):
        measurement_records.append({
            'eventID': event_id,
            'measurementType': 'wind direction',
            'measurementValue': str(row['wind_dir']),
            'measurementUnit': ''
        })
    
    # Wave height
    if pd.notna(row['wave_height']):
        measurement_records.append({
            'eventID': event_id,
            'measurementType': 'wave height',
            'measurementValue': str(row['wave_height']),
            'measurementUnit': ''
        })
    
    # Cloud cover
    if pd.notna(row['cloud_cover_10th']):
        measurement_records.append({
            'eventID': event_id,
            'measurementType': 'cloud cover',
            'measurementValue': str(row['cloud_cover_10th']),
            'measurementUnit': 'tenths'
        })
    
    # Ropeless gear ID
    if pd.notna(row['ropeless_id']):
        measurement_records.append({
            'eventID': event_id,
            'measurementType': 'ropeless gear ID',
            'measurementValue': str(row['ropeless_id']),
            'measurementUnit': ''
        })
    
    # Cruise ID as measurement
    # Schema slot: cruise_id -> dwc:measurementValue (transformation_type: pivot_to_mof)
    if pd.notna(row['cruise_id']):
        measurement_records.append({
            'eventID': event_id,
            'measurementType': 'cruise ID',
            'measurementValue': str(int(row['cruise_id'])),
            'measurementUnit': ''
        })

# Gear measurements from CPUE
for idx, row in cpue_df.iterrows():
    event_id = row['Station']
    
    # Pot position
    if pd.notna(row['Pot_position']):
        measurement_records.append({
            'eventID': event_id,
            'measurementType': 'pot position',
            'measurementValue': str(row['Pot_position']),
            'measurementUnit': ''
        })
    
    # Pot ID
    if pd.notna(row['Pot_ID']):
        measurement_records.append({
            'eventID': event_id,
            'measurementType': 'pot ID',
            'measurementValue': str(row['Pot_ID']),
            'measurementUnit': ''
        })
    
    # Distance category (near/far)
    if pd.notna(row['Near_Far']):
        measurement_records.append({
            'eventID': event_id,
            'measurementType': 'distance category',
            'measurementValue': str(row['Near_Far']),
            'measurementUnit': ''
        })

# Distance category from Measurements
for idx, row in measurements_df.iterrows():
    if pd.notna(row['Near/Far']):
        measurement_records.append({
            'eventID': row['Station'],
            'measurementType': 'distance category',
            'measurementValue': str(row['Near/Far']),
            'measurementUnit': ''
        })

measurement_or_fact = pd.DataFrame(measurement_records)

print(f"\nCreated MeasurementOrFact extension with {len(measurement_or_fact)} measurements")

# ============================================================================
# Write outputs
# ============================================================================

event_core.to_csv('outputs/py_dwc_event.csv', index=False, na_rep='')
occurrence_combined.to_csv('outputs/py_dwc_occurrence.csv', index=False, na_rep='')
measurement_or_fact.to_csv('outputs/py_dwc_measurementorfact.csv', index=False, na_rep='')
